### ISSUE TITLE
Feature request: Ability to pass objects to @Command annotated callables.

### ISSUE THREAD

**Issue by Flightkick** [remkop/picocli#1016]
Maybe I'm missing existing functionality but I couldn't find in the docs nor in the issues here at Github.

I would like to pass context (non-user input) to commands being executed.    
In this case, I would like to pass an event with the `CommandLine.execute(String.. args)` which I want to use in the `call()` method of the command.
[Relevant documentation](https://picocli.info/#execute)

I would like to achieve something like this:
```java
@Command(name = "myapp", mixinStandardHelpOptions = true, version = "1.0")
class MyApp implements Callable<Integer> {

    @Option(names = "-x") int x;

    @Context(name = "event")                        // <----
    Event event;

    @Override
    public Integer call() { // business logic
        event.broadcast("something");               // <----
        System.out.printf("x=%s%n", x);
        return 123; // exit code
    }

    public static void main(String... args) { // bootstrap the application
        System.exit(new CommandLine(new MyApp()).execute(args));
    }
}
```

And then execute the command in the following way:
```java
cmd = new CommandLine(...)
...
cmd.setContext("event", event);
cmd.execute("myapp");
```

**Comment by remkop** [comment]
Hello @Flightkick, thank you for raising this.

I could be wrong, but the example you showed looks a bit like field injection, which could be accomplished with Spring, or some other Dependency Injection container. 

If that is what you are looking for, please take a look at the dependency injection section of the manual: https://picocli.info/#_dependency_injection

**Comment by Flightkick** [comment]
That was fast, thanks!

Normally dependency injection would suffice but in this case the Event object is context specific to that .execute() call only. It's different for every call.

An event is triggered. The event includes the command query along with it's parameters and options but also the context on how to reply to the sender. The query is then parsed and executed and the event is needed in order to reply.

**Comment by remkop** [comment]
Ok, from your description, the parsing and execution only happens when an event is received.
So there is some callback that provides the event object and that is when the parsing should happen?

How about something like this then:

```java
@Command(name = "myapp", mixinStandardHelpOptions = true, version = "1.0")
class MyApp implements Callable<Integer> {

    @Option(names = "-x") int x;

    Event event;

    MyApp(Event event) {
        this.event = event;
    }

    @Override
    public Integer call() { // business logic
        event.broadcast("something");  
        System.out.printf("x=%s%n", x);
        return 123; // exit code
    }

    static String[] args; // ?
    public void onEvent(Event event) {
        System.exit(new CommandLine(new MyApp(event)).execute(args));
    }

    public static void main(String... args) {
        MyApp.args = args;
        // start the event loop...
    }
}
```

**Comment by Flightkick** [comment]
> Ok, from your description, the parsing and execution only happens when an event is received.
So there is some callback that provides the event object and that is when the parsing should happen?

Yes, that's correct @remkop 

There's one implementation specific catch here. I'm instantiating the commands through Spring Boot and make them register themselves at the command handler. This way adding commands is dynamic.

I've found a way to achieve what I want to using reflection.

My commands extend an abstract Command as shown below. Spring Boot injects the CommandHandler and post-construct this abstract will make sure the command registers itself with the command handler.
```java
@Slf4j
@RequiredArgsConstructor
public abstract class Command implements Callable<Integer> {

	private final CommandHandler commandHandler;

        @Setter
	protected Event event;

	@PostConstruct
	public void init() {
                // Called by Spring Boot after initialization of component.
		commandHandler.register(this);
	}

	@Override
	public abstract Integer call();

}
```

Then in the command handler, every time an event is received instantiate new Commands and set the context event.
```java
@Slf4j
@Component
public class CommandHandler extends MyEventListener {

	private final CommandLine cmd;

        private final List<Class<? extends Command>> commands = new CopyOnWriteArrayList<>();


	public void register(final Command command) {
                // Add command class to a list so we can instantiate it later.
                final Class<? extends Command> clazz = command.getClass();
		log.info("Registering command {}", clazz.getCanonicalName());
		commands.add(clazz);
	}

	@Override
	public void onMessageReceived(final Event event) {
        // Instantiate a new CommandLine
        final CommandLine comm = new CommandLine(new MainCommand());
        // For each registered command, use reflection to instantiate and set the context
	commands.forEach(command -> {
		try {
			final Constructor<? extends Command> ctor = command.getDeclaredConstructor(this.getClass());
			final Command c = ctor.newInstance(this);
			c.setEvent(event);
			comm.addSubcommand(c);
		} catch (final NoSuchMethodException e) {
			log.error("Failed to find the command constructor", e);
		} catch (final IllegalAccessException e) {
			log.error("Illegal Access to command", e);
		} catch (final InstantiationException e) {
			log.error("Failed to instantiate command", e);
		} catch (final InvocationTargetException e) {
			log.error("Failed to invoke target command", e);
		}
	});
        // Execute the command
	cmd.execute(event.getCommandQuery());
    }

}
```

So now my command classes have access to the context of the event:
```java
@Slf4j
@CommandLine.Command(name = "hello",
                     aliases = {"hi"},
                     description = "Simple Hello World command",
                     subcommands = {CommandLine.HelpCommand.class})
@Component
public class Hello extends Command {

	public Hello(final CommandHandler commandHandler) {
		super(commandHandler);
	}

	@Override
	public Integer call() {
		event.reply("something");  // Access to event context!
                log.info("Responded to hello command!")
		return 0;
	}

}
```

It might be nice if picocli is able to handle this reflection part somehow.

**Comment by remkop** [comment]
Perhaps use a custom factory in `onMessageReceived`? https://picocli.info/#_custom_factory

One idea is to extend (or copy the source and create your own version of) the PicocliSpringFactory: https://github.com/remkop/picocli/tree/master/picocli-spring-boot-starter

**Comment by Flightkick** [comment]
Thanks! I was able to implement a custom factory to achieve this. My request was based on the assumption I could re-use the CommandLine. This works fine too :)

**Comment by remkop** [comment]
Glad to hear that!
I’ll close this ticket; feel free to open another one if you have more questions.

**Comment by victorchicu** [comment]
Hi Flightkick,
Could you please share an example of code how you implement it via a custom factory?
My use case is pretty similar to your I need to pass in run or call method not only arguments but an additional object (let's say metadata per call)
Thanks, with best regards Victor

**Comment by victorchicu** [comment]
I found a way how to pass objects to commands the only downside is that I have to create CommandLine object each time I parse or execute user input.
First of all I'm using picocli-spring-boot-starter. 
To inject beans into the command via constructor you have to initialize PicocliSpringFactory.

**Example**:
```java
 CommandLine commandLine = new CommandLine(new BotCommand(update), new PicocliSpringFactory(context));
 commandLine.setExecutionStrategy(new CommandLine.RunLast());
```

When you instantiate the command line you could inject whatever you want to ParentCommand in my case is BotCommand then from child command you could hook @ParentCommand and access your objects.  

**Example**:
```java
@Component
@CommandLine.Command(
        name = "track"
)
public class TrackCommand extends BaseCommand {
    @CommandLine.ParentCommand
    private BotCommand botCommand;

    @Override
    public void run() {
       Update p = botCommand.getUpdate();
    }
 }
````
I don't really like this approach because I have to create a lot of CommandLine objects but at least it works.
If you have a more elegant solution I would more than happy :)

**Comment by Flightkick** [comment]
@victorchicu Due to other responsibilities, I had to shelve the project I was working on at the time, but I had a similar use-case and solution. My goal was to allow extensibility for creators. Allowing them to add their own commands as well. I ended up adding custom annotations as a facade to the underlying PicoCli engine. Works well but I can't recommend the approach as it will involve a lot of reflections usage, which doesn't necessarily make it a great thing to maintain. And it doesn't solve the issue you're facing with multiple CommandLine objects.

> @victorchicu: I don't really like this approach because I have to create a lot of CommandLine objects but at least it works.

That's what I bumped into as well, trying to re-use the CommandLine but that somehow wouldn't work. (long time ago, forgot why exactly), hence my response:

> @Flightkick My request was based on the assumption I could re-use the CommandLine. This works fine too :)

Good luck with the project 😉
